lib/puppet/type.rb

```diff
@@ -2117,8 +2117,17 @@ def autorelation(rel_type, rel_catalog = nil)
     auto_rel = "eachauto#{rel_type}".to_sym
 
     self.class.send(auto_rel) { |type, block|
-      # Ignore any types we can't find, although that would be a bit odd.
-      next unless Puppet::Type.type(type)
+      if not Puppet::Type.type(type)
+        # We know it is not a type at this point, let's see if it exists as a defined type in the catalog
+        # We split and join on :: because defined types will likely have colon separators
+        # All defined types are of type :component
+        found_defines = rel_catalog.resources.select do |x|
+          x.type == :component && x.title_class == type.to_s.split('::').map(&:capitalize).join('::')
```

Comment:
Further discussion in PR Triage:
- Determining whether a resource is a defined_type should be a query to the resource. Even if we have to use a heuristic on the name, it seems like we should encapsulate that in the resource.
- It seems like somehow https://github.com/puppetlabs/puppet/blob/master/lib/puppet/resource/type.rb#L25 should be accessible and usable in this instance.